tokio/src/io/async_fd.rs

```diff
@@ -508,6 +508,39 @@ impl<T: AsRawFd> AsyncFd<T> {
     pub async fn writable_mut<'a>(&'a mut self) -> io::Result<AsyncFdReadyMutGuard<'a, T>> {
         self.readiness_mut(Interest::WRITABLE).await
     }
+
+    /// Reads or writes from the file descriptor using a user-provided IO operation.
+    ///
+    /// The readiness of the file descriptor is awaited and when the file descriptor is ready,
+    /// the provided closure is called. The closure should attempt to perform
+    /// IO operation on the file descriptor by manually calling the appropriate syscall.
+    /// If the operation fails because the file descriptor is not actually ready,
+    /// then the closure should return a `WouldBlock` error. In such case the
+    /// readiness flag is cleared and the file descriptor readiness is awaited again.
+    /// This loop is repeated until the closure returns an `Ok` or an error
+    /// other than `WouldBlock`.
+    ///
+    /// The closure should only return a `WouldBlock` error if it has performed
+    /// an IO operation on the file descriptor that failed due to the file descriptor not being
+    /// ready. Returning a `WouldBlock` error in any other situation will
+    /// incorrectly clear the readiness flag, which can cause the file descriptor to
+    /// behave incorrectly.
+    ///
+    /// The closure should not perform the IO operation using any of the methods
+    /// defined on the Tokio `AsyncFd` type, as this will mess with the
+    /// readiness flag and can cause the file descriptor to behave incorrectly.
+    ///
+    /// This method is not intended to be used with combined interests.
+    /// The closure should perform only one type of IO operation, so it should not
+    /// require more than one ready state. This method may panic or sleep forever
+    /// if it is called with a combined interest.
+    pub async fn async_io<R>(
```

Comment:
Can you include an example of the equivalent code using the guards?

Reply:
This is the code we currently have: 

```rust
    let output = loop {
        let mut guard = async_fd.writable().await?;
        match guard.try_io(|inner| inner.get_ref().send(buf)) {
            Ok(result) => 
                break result;
            },
            Err(_would_block) => {
                continue;
            }
        }
    };
``` 

with this addition, we can instead write:

```rust
let output = async_fd.async_io(Interest::WRITABLE, || async_fd.get_ref().send(buf)).await;
```

Would you like to see this in the documentation comment as well?

<hr>

I don't think the guard is relevant here because what it does is already taken care of by the underlying `Interest::async_io`. The `try_io` method on `AsyncFdReadyGuard` will clear the ready flag if the closure returns WouldBlock

```rust
pub fn try_io<R>(
    &mut self,
    f: impl FnOnce(&'a AsyncFd<Inner>) -> io::Result<R>,
) -> Result<io::Result<R>, TryIoError> {
    let result = f(self.async_fd);

    if let Err(e) = result.as_ref() {
        if e.kind() == io::ErrorKind::WouldBlock {
            self.clear_ready();
        }
    }

    match result {
        Err(err) if err.kind() == io::ErrorKind::WouldBlock => Err(TryIoError(())),
        result => Ok(result),
    }
}
```

but for the `AsyncFd::async_io` implemented here, that is not needed because the underlying `Interest::async_io` already clears the flag:


```rust
pub(crate) async fn async_io<R>(&self, interest: Interest, mut f: impl FnMut() -> io::Result<R>) -> io::Result<R> {
    loop {
        let event = self.readiness(interest).await?;

        match f() {
            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                self.clear_readiness(event);
            }
            x => return x,
        }
    }
}
```

Reply:
Yeah, what you mentioned at the top would be a good addition to the documentation. I'm not suggesting that you change the implementation.

Reply:
done. I also turned various mentions of types/variants into links since that is what the rest of the module seems to do.